### PR TITLE
impl traits needed to deserialize from event_content_collection macro

### PR DESCRIPTION
**Pull request by DevinR528**
Move the EventContent and content marker traits into macro.
Move the raw AnyStateEventContent enum and RawEventContent trait into macro, generate all of the AnyStateEventContent type and impls needed for raw.

### CODE REVIEW

**Comment by jplatte** [comment]
Re. the private documentation errors, you can add `#![allow(clippy::missing_docs_in_private_items)]` to the macro crate's crate attributes. I don't think we need to document every private function.

File: ruma-events-macros/src/collection.rs [hunk 1]
```diff
@@ -10,52 +10,173 @@ use parse::RumaCollectionInput;
...
+                #[doc = #event_type_str]
+                #variants(#raw_content)
+            ),*
+        }
+    };
+
+    let raw_event_content_impl = quote! {
+        impl ::ruma_events::RawEventContent for #ident {
+            fn from_parts(event_type: &str, content: Box<::serde_json::value::RawValue>) -> Result<Self, String> {
+                fn deserialize_variant<T: ::ruma_events::RawEventContent>(
```
Comment (jplatte) [inline ruma-events-macros/src/collection.rs:106]: This `quote!` invocation would be shorter and possibly more readable if it didn't have this local function and instead had the same code inside the repetition. I'm not entirely certain though. What do you think?

File: ruma-events-macros/src/collection.rs [hunk 2]
```diff
@@ -10,52 +10,173 @@ use parse::RumaCollectionInput;
...
+            fn event_type(&self) -> &str {
+                match self {
+                    #( Self::#variants(content) => content.event_type()),*
+                }
+            }
+        }
+    };
+
+    let try_from_raw_impl = quote! {
+        impl ::ruma_events::FromRaw for #ident {
```
[the remark this answers is not in the document]
Comment (DevinR528) [reply ruma-events-macros/src/collection.rs:46]: Is the error type `std::convert::Infallible`?

File: ruma-events-macros/src/collection.rs (hunk 1 above)
Comment (DevinR528) [reply ruma-events-macros/src/collection.rs:106]: Yea the type signature makes it really complicated looking.

File: ruma-events-macros/src/collection.rs [hunk 3]
```diff
@@ -43,14 +42,28 @@ pub fn expand_collection(input: RumaCollectionInput) -> syn::Result<TokenStream>
...
                 match raw {
-                    #( #variants(c) => Self::#variants(::ruma_events::FromRaw::from_raw(c))),*
+                    #( #variants(c) => Ok(Self::#variants(::ruma_events::TryFromRaw::try_from_raw(c)?))),*
+                }
+            }
+        }
+    };
+
+    let serialize_impl = quote! {
+        impl ::serde::Serialize for #ident {
```
Comment (jplatte) [inline ruma-events-macros/src/collection.rs:60]: I wonder whether we need this: Shouldn't `#[serde(untagged)]` make serde's `Serialize` derive generate the same code?

Comment (DevinR528) [reply ruma-events-macros/src/collection.rs:60]: Yup it did, I'll go back to that :)

File: ruma-events-macros/src/collection.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply ruma-events-macros/src/collection.rs]: Oh, huh. Well no, that's not going to work for all event types. I'll have to go check what `master`s enums do.

[the remark this answers is not in the document]
Comment (jplatte) [reply ruma-events-macros/src/collection.rs]: Ah, they use `String`. Probably good to just replicate that: https://github.com/ruma/ruma-events/blob/master/src/collections/only.rs#L168

File: ruma-events-macros/src/collection.rs [hunk 4]
```diff
@@ -45,13 +45,21 @@ pub fn expand_collection(input: RumaCollectionInput) -> syn::Result<TokenStream>
...
+            type Err = String;
 
             fn try_from_raw(raw: Self::Raw) -> Result<Self, Self::Err> {
                 use raw::#ident::*;
 
                 match raw {
-                    #( #variants(c) => Ok(Self::#variants(::ruma_events::TryFromRaw::try_from_raw(c)?))),*
+                    #( #variants(c) => {
+                            let content = ::ruma_events::TryFromRaw::try_from_raw(c)
+                                .map_err(|e: <#content as ::ruma_events::TryFromRaw>::Err| e.to_string());
```
Comment (jplatte) [inline ruma-events-macros/src/collection.rs:56]:
```suggestion
                                .map_err(|e: <#content as ::ruma_events::TryFromRaw>::Err| e.to_string())?;
```

File: ruma-events-macros/src/collection.rs [hunk 5]
```diff
@@ -45,13 +45,21 @@ pub fn expand_collection(input: RumaCollectionInput) -> syn::Result<TokenStream>
...
+                                // without this ^^^^^^^^^^^ the compiler fails to infer the type
+                            Ok(Self::#variants(
+                                content?
+                            ))
```
Comment (jplatte) [inline ruma-events-macros/src/collection.rs:60]:
```suggestion
                            Ok(Self::#variants(content))
```

**Pull request merged**